# Preprocessing for PubMed Dataset

## 1. Load Dataset

In [3]:
# load by huggingface datasets load_dataset
from datasets import load_dataset
pubmed_copy = load_dataset('json', data_dir='../datasets/pubmed-dataset-copy')

Found cached dataset json (/home/is/kaifan-l/.cache/huggingface/datasets/json/default-4860f0258abb954c/0.0.0/8bb11242116d547c741b2e8a1f18598ffdd40a1d4f2a2872c7a28b697434bc96)


  0%|          | 0/3 [00:00<?, ?it/s]

In [37]:
# load by pandas
import pandas as pd
train_df = pd.read_json('../datasets/pubmed-dataset-copy/train.json', lines=True)
val_df = pd.read_json('../datasets/pubmed-dataset-copy/val.json', lines=True)
test_df = pd.read_json('../datasets/pubmed-dataset-copy/test.json', lines=True)

In [40]:
train_df

article_id                                       article_text  \
0       PMC3872579  [a recent systematic analysis showed that in 2...   
1       PMC3770628  [it occurs in more than 50% of patients and ma...   
2       PMC5330001  [tardive dystonia ( td ) , a rarer side effect...   
3       PMC4386667  [lepidoptera include agricultural pests that ,...   
4       PMC4307954  [syncope is caused by transient diffuse cerebr...   
...            ...                                                ...   
119919  PMC3502213  [eukaryotic cells depend on vesicle - mediated...   
119920  PMC3198562  [as regards the selection criteria of the post...   
119921  PMC4436536  [in most of the peer review publications in th...   
119922  PMC4251613  [the reveal registry is a longitudinal registr...   
119923  PMC3389695  [cerebral palsy is a nonprogressive central ne...   

                                            abstract_text  labels  \
0       [<S> background : the present study was carrie...     NaN   
1       [<S> backgroundanemia in patients with cancer ...     NaN   
2       [<S> tardive dystonia ( td ) is a serious side...     NaN   
3       [<S> many lepidopteran insects are agricultura...     NaN   
4       [<S> we present an unusual case of recurrent c...     NaN   
...                                                   ...     ...   
119919  [<S> long - distance trafficking of membranous...     NaN   
119920  [<S> aims and objectives : to study the stress...     NaN   
119921  [<S> abstractbackgroundthe objective of this s...     NaN   
119922  [<S> background : patients with pulmonary arte...     NaN   
119923  [<S> \n objective . to determine </S>, <S> if ...     NaN   

                                            section_names  \
0       [INTRODUCTION, MATERIALS AND METHODS, Particip...   
1       [Introduction, Patients and methods, Study des...   
2       [INTRODUCTION, CASE REPORT, DISCUSSION, Declar...   
3       [1. Introduction, 2. Insect Immunity, 3. Signa...   
4       [Introduction, Case report, Discussion, Confli...   
...                                                   ...   
119919  [Introduction, Motor-Dependent Transport of Ra...   
119920  [INTRODUCTION, MATERIALS AND METHODS, Modeling...   
119921  [Introduction, Methods, Results, Discussion, L...   
119922  [TRIAL REGISTRY:, Materials and Methods, REVEA...   
119923  [1. Introduction, 2. Method, 3. Results, 4. Di...   

                                                 sections  
0       [[a recent systematic analysis showed that in ...  
1       [[it occurs in more than 50% of patients and m...  
2       [[tardive dystonia ( td ) , a rarer side effec...  
3       [[lepidoptera include agricultural pests that ...  
4       [[syncope is caused by transient diffuse cereb...  
...                                                   ...  
119919  [[eukaryotic cells depend on vesicle - mediate...  
119920  [[fiber post systems are routinely used in res...  
119921  [[in most of the peer review publications in t...  
119922  [[], [the reveal registry is a longitudinal re...  
119923  [[cerebral palsy is a nonprogressive central n...  

[119924 rows x 6 columns]

In [41]:
val_df

article_id                                       article_text  \
0     PMC4810892  [approximately , one - third of patients with ...   
1     PMC4852153  [there is an epidemic of stroke in low and mid...   
2     PMC5307612  [\n cardiovascular diseases account for the hi...   
3     PMC3309663  [results of a liquid culturing system ( bd bac...   
4     PMC4413883  [the need for magnetic resonance imaging ( mri...   
...          ...                                                ...   
6628  PMC3934811  [a 48-year - old woman with a positive brca1 g...   
6629  PMC4503916  [with an increasing number of laparoscopic cho...   
6630  PMC4754593  [upton and mccomas first reported double crush...   
6631  PMC5317293  [acute respiratory distress syndrome ( ards ) ...   
6632  PMC4672074  [cardiac resynchronization therapy ( crt ) is ...   

                                          abstract_text  labels  \
0     [<S> background and aim : there is lack of sub...     NaN   
1     [<S> backgroundthe questionnaire for verifying...     NaN   
2     [<S> \n background : timely access to cardiova...     NaN   
3     [<S> to determine differences in the ability o...     NaN   
4     [<S> aimsour aim was to evaluate the potential...     NaN   
...                                                 ...     ...   
6628  [<S> a 48-year - old woman with a positive brc...     NaN   
6629  [<S> we present a case on abdominal wall absce...     NaN   
6630  [<S> double compression of the ulnar nerve , i...     NaN   
6631  [<S> backgroundno definitive conclusions have ...     NaN   
6632  [<S> backgroundvarious difficulties can occur ...     NaN   

                                          section_names  \
0     [Introduction, Subjects and Methods, Results, ...   
1     [1. Introduction, 2. Methods, 2.1. Study sites...   
2     [Introduction, Methods, Results, Discussion, C...   
3      [The Study, Conclusions, Supplementary Material]   
4     [Introduction, Methods, Patient selection, Mag...   
...                                                 ...   
6628              [Case Report, Discussion, Conclusion]   
6629  [INTRODUCTION, CASE REPORT, DISCUSSION, CONFLI...   
6630            [INTRODUCTION, CASE REPORT, DISCUSSION]   
6631  [Background, Material and Methods, Patient sel...   
6632  [Introduction, Material and methods, Study pat...   

                                               sections  
0     [[venous thromboembolism ( vte ) comprising of...  
1     [[there is an epidemic of stroke in low and mi...  
2     [[\n cardiovascular diseases account for the h...  
3     [[results of a liquid culturing system ( bd ba...  
4     [[the need for magnetic resonance imaging ( mr...  
...                                                 ...  
6628  [[a 48-year - old woman with a positive brca1 ...  
6629  [[with an increasing number of laparoscopic ch...  
6630  [[upton and mccomas first reported double crus...  
6631  [[acute respiratory distress syndrome ( ards )...  
6632  [[cardiac resynchronization therapy ( crt ) is...  

[6633 rows x 6 columns]

In [42]:
test_df

article_id                                       article_text  \
0     PMC5075302  [anxiety affects quality of life in those livi...   
1     PMC3309138  [small non - coding rnas are transcribed into ...   
2     PMC4086000  [ohss is a serious complication of ovulation i...   
3     PMC3603086  [congenital adrenal hyperplasia ( cah ) refers...   
4     PMC4414990  [type 1 diabetes ( t1d ) results from the dest...   
...          ...                                                ...   
6653  PMC2817809  [neurocysticercosis is the most common parasit...   
6654  PMC5421177  [most chd prevalence data are based on populat...   
6655  PMC3336056  [renal transplantation rates are low among pat...   
6656  PMC4461384  [acute kidney injury ( aki ) , defined as an a...   
6657  PMC4637496  [human brain plasticity or neuroplasticity ref...   

                                          abstract_text  labels  \
0     [<S> research on the implications of anxiety i...     NaN   
1     [<S> small non - coding rnas include sirna , m...     NaN   
2     [<S> objective : to evaluate the efficacy and ...     NaN   
3     [<S> congenital adrenal hyperplasia is a group...     NaN   
4     [<S> objective(s):pentoxifylline is an immunom...     NaN   
...                                                 ...     ...   
6653  [<S> cysticercosis is an infection with larval...     NaN   
6654  [<S> background : congenital heart disease ( c...     NaN   
6655  [<S> we describe a successful living donor liv...     NaN   
6656  [<S> objective : the aim was to investigate th...     NaN   
6657  [<S> accumulating evidence has indicated that ...     NaN   

                                          section_names  \
0     [1. Introduction, 2. Methods, 3. Results, 4. D...   
1     [Introduction, Aberrant Expression of miRNA in...   
2     [Introduction, Materials and Methods, Results,...   
3                                          [I, M, R, D]   
4     [Introduction, Materials and Methods, Drug and...   
...                                                 ...   
6653  [1. Introduction, 2. Materials and Methods, 3....   
6654  [I, M, Subjects, Ethical approval, Protocol, D...   
6655            [Introduction, Case report, Discussion]   
6656  [INTRODUCTION, METHODS, Search Strategy, Study...   
6657  [1. Introduction, 2. Materials and Methods, 3....   

                                               sections  
0     [[anxiety affects quality of life in those liv...  
1     [[small non - coding rnas are transcribed into...  
2     [[ohss is a serious complication of ovulation ...  
3     [[congenital adrenal hyperplasia ( cah ) refer...  
4     [[type 1 diabetes ( t1d ) results from the des...  
...                                                 ...  
6653  [[neurocysticercosis is the most common parasi...  
6654  [[in developing countries , congenital heart d...  
6655  [[renal transplantation rates are low among pa...  
6656  [[acute kidney injury ( aki ) , defined as an ...  
6657  [[human brain plasticity or neuroplasticity re...  

[6658 rows x 6 columns]

In [44]:
test_df['sections'][0]

[["anxiety affects quality of life in those living with parkinson 's disease ( pd ) more so than overall cognitive status , motor deficits , apathy , and depression [ 13 ] .",
  'although anxiety and depression are often related and coexist in pd patients , recent research suggests that anxiety rather than depression is the most prominent and prevalent mood disorder in pd [ 5 , 6 ] . yet ,',
  'our current understanding of anxiety and its impact on cognition in pd , as well as its neural basis and best treatment practices , remains meager and lags far behind that of depression .',
  'overall , neuropsychiatric symptoms in pd have been shown to be negatively associated with cognitive performance .',
  'for example , higher depression scores have been correlated with lower scores on the mini - mental state exam ( mmse ) [ 8 , 9 ] as well as tests of memory and executive functions ( e.g. , attention ) [ 1014 ] .',
  'likewise , apathy and anhedonia in pd patients have been associated with

* Pubmed statistic:
  | Split | samples | 
  | --- | --- |
  | Train | 119924 | 
  | Val | 6633 |
  | Test | 6658 |

* a sample of vanilla dataset
  * labels seems to used for classification task, the labels of vanilla datasets are all None, i didn't do any processing.
  
  | article_id | article_text | abstract_text | labels | section_names | sections |
  | --- | --- | --- | --- | --- | --- |
  | str | List[str] | List[str] | None | List[str] | List[List[str]] |
  | 'PMC3872579' | split by sentence<br>(似乎按照句子进行分割了, 如果要使用需要进行join) | ['<BOS> background : the present study was carried out to assess the effects of community nutrition intervention based on advocacy approach on malnutrition status among school - aged children in shiraz , iran.materials and methods : this case - control nutritional intervention has been done between 2008 and 2009 on 2897 primary and secondary school boys and girls ( 7 - 13 years old ) based on advocacy approach in shiraz , iran . </EOS>', <br> '<BOS> the project provided nutritious snacks in public schools over a 2-year period along with advocacy oriented actions in order to implement and promote nutritional intervention . for evaluation of effectiveness of the intervention growth monitoring indices of pre- and post - intervention were statistically compared.results:the frequency of subjects with body mass index lower than 5% decreased significantly after intervention among girls ( p = 0.02 ) . </EOS>', <br> '<BOS> however , there were no significant changes among boys or total population . </EOS>', <br> '<BOS> the mean of all anthropometric indices changed significantly after intervention both among girls and boys as well as in total population . </EOS>', <br> "<BOS> the pre- and post - test education assessment in both groups showed that the student 's average knowledge score has been significantly increased from 12.5  3.2 to 16.8  4.3 ( p < 0.0001).conclusion : this study demonstrates the potential success and scalability of school feeding programs in iran . </EOS>", <br> '<BOS> community nutrition intervention based on the advocacy process model is effective on reducing the prevalence of underweight specifically among female school aged children . </EOS>'] <br> 分割方式不明, 需要进行join并重新分割 | None | ['INTRODUCTION', <br>'MATERIALS AND METHODS', <br>'Participants'Instruments', <br>'Procedure', <br>'First step', <br>'Second step', <br>'Third step', <br>'Forth step', <br>'Interventions', <br>'Fifth step (assessment)', <br>'Data analysis', <br>'RESULTS', <br>'DISCUSSION', <br>'CONCLUSION'] <br><br>请注意METHODS包含了从Participants'Instruments到Data analysis的部分 | [<br>section[seq1, seq2, ...], <br>section[seq1, seq2, ...], <br>...] |
  | reformat之后的格式 | --- | --- | --- | --- | --- |
  | | | | | | List[str]|


## 2. Dataset Preprocessing

### 2.1 Keyword Table
The summary(abstract) of pubmed dataset have not been split into multiple pieces, but luckily, it has section strutures so we could mannually process it by keyword matching.
> reference: [GovReport](https://arxiv.org/pdf/2104.02112v2.pdf)
> see page 18

In [6]:
segmentation_keyword_table = {
    # Introduction and Literature
    'part_1': ['introduction', 'case', 'objectives', 'purposes', 
               'objective', 'purpose', 'background', 'literature',],
    
    # Methods
    'part_2': ['materials and methods', 'techniques', 'methodology',
               'materials', 'research design', ' study design'],
    
    # Results
    'part_3': ['results', 'experiments', 'observations'],
    
    # Discussion and Conlusion
    'part_4': ['discussion', 'limitation', 'conclusions', 'concluding']
    
}

## 2.2 Sections

### 2.2.1 Concat sentences into one setence <br>(每一行是一个完整的sections, 由多个section组成, 其中每个section有多个句子, 需要进行合并)

In [50]:
pubmed_copy['train'][0]['sections']

[['a recent systematic analysis showed that in 2011 , 314 ( 296 - 331 ) million children younger than 5 years were mildly , moderately or severely stunted and 258 ( 240 - 274 ) million were mildly , moderately or severely underweight in the developing countries .',
  'in iran a study among 752 high school girls in sistan and baluchestan showed prevalence of 16.2% , 8.6% and 1.5% , for underweight , overweight and obesity , respectively .',
  'the prevalence of malnutrition among elementary school aged children in tehran varied from 6% to 16% .',
  'anthropometric study of elementary school students in shiraz revealed that 16% of them suffer from malnutrition and low body weight .',
  'snack should have 300 - 400 kcal energy and could provide 5 - 10 g of protein / day . nowadays , school nutrition programs are running as the national programs , world - wide . national school lunch program in the united states',
  'there are also some reports regarding school feeding programs in developi

In [45]:
train_df['sections'][0]

[['a recent systematic analysis showed that in 2011 , 314 ( 296 - 331 ) million children younger than 5 years were mildly , moderately or severely stunted and 258 ( 240 - 274 ) million were mildly , moderately or severely underweight in the developing countries .',
  'in iran a study among 752 high school girls in sistan and baluchestan showed prevalence of 16.2% , 8.6% and 1.5% , for underweight , overweight and obesity , respectively .',
  'the prevalence of malnutrition among elementary school aged children in tehran varied from 6% to 16% .',
  'anthropometric study of elementary school students in shiraz revealed that 16% of them suffer from malnutrition and low body weight .',
  'snack should have 300 - 400 kcal energy and could provide 5 - 10 g of protein / day . nowadays , school nutrition programs are running as the national programs , world - wide . national school lunch program in the united states',
  'there are also some reports regarding school feeding programs in developi

In [47]:
def sec_join(df):
    # 每一行
    sample_list = []
    for sample in df['sections']:
        # 每一个sample的每一个section
        sections = []
        for section in sample: 
            temp = " ".join(section)
            sections.append(temp)
        
        sample_list.append(sections)
    
    return sample_list

In [49]:
train_df['sections'] = sec_join(train_df)
test_df['sections'] = sec_join(test_df)
val_df['sections'] = sec_join(val_df)

["a recent systematic analysis showed that in 2011 , 314 ( 296 - 331 ) million children younger than 5 years were mildly , moderately or severely stunted and 258 ( 240 - 274 ) million were mildly , moderately or severely underweight in the developing countries . in iran a study among 752 high school girls in sistan and baluchestan showed prevalence of 16.2% , 8.6% and 1.5% , for underweight , overweight and obesity , respectively . the prevalence of malnutrition among elementary school aged children in tehran varied from 6% to 16% . anthropometric study of elementary school students in shiraz revealed that 16% of them suffer from malnutrition and low body weight . snack should have 300 - 400 kcal energy and could provide 5 - 10 g of protein / day . nowadays , school nutrition programs are running as the national programs , world - wide . national school lunch program in the united states there are also some reports regarding school feeding programs in developing countries . in vietnam 

In [72]:
train_df.to_json('train.json', orient='records', lines=True)
test_df.to_json('test.json', orient='records', lines=True)
val_df.to_json('val.json', orient='records', lines=True)

### 2.2.2 删除空行（包括None）